cpp/src/fil/infer.cu

```diff
@@ -533,45 +544,61 @@ template <int NITEMS, leaf_algo_t leaf_algo, bool cols_in_shmem,
 __global__ void infer_k(storage_type forest, predict_params params) {
   extern __shared__ char smem[];
   float* sdata = (float*)smem;
+  int sdata_stride = params.sdata_stride();
+  int rows_per_block = NITEMS << params.log2_threads_per_tree;
   int num_cols = params.num_cols;
-  for (size_t block_row0 = blockIdx.x * NITEMS; block_row0 < params.num_rows;
-       block_row0 += NITEMS * gridDim.x) {
-    size_t num_input_rows = min((size_t)NITEMS, params.num_rows - block_row0);
+  for (size_t block_row0 = blockIdx.x * rows_per_block;
+       block_row0 < params.num_rows; block_row0 += rows_per_block * gridDim.x) {
+    int block_num_rows =
+      min((size_t)rows_per_block, params.num_rows - block_row0);
     const float* block_input = params.data + block_row0 * num_cols;
     if (cols_in_shmem) {
       // cache the row for all threads to reuse
       size_t feature = 0;
 #pragma unroll
-      for (feature = threadIdx.x; feature < num_input_rows * num_cols;
-           feature += blockDim.x)
-        sdata[feature] = block_input[feature];
+      for (feature = threadIdx.x; feature < block_num_rows * num_cols;
+           feature += blockDim.x) {
+        int row = feature / num_cols, col = feature % num_cols;
```

Comment:
Would using the "division by multiplication" optimization speed it up? It should be available in cuML or RAFT somewhere.

Reply:
I'd prefer to benchmark it, to avoid overcomplicating code with premature optimizations and to keep it readable

Reply:
there can be some speedup in an anomalous edge case. In most cases, I expect <1% benefit, however, the optimization is very simple in the code. We're still determining the best way to represent this optimization

Reply:
in the common case, this would actually result in a slowdown, since usually `num_cols < 256 == FIL_TPB`. Most of the 5% gain in the anomalous case with e.g. 4000 columns and 1 tree of depth 1 can be made by switching the index type, which I did.
I added an assert to make sure FIL never quietly overflows. This assert would only trigger if a future SM had >262KB of shmem, at which point we can decide what to do.